### PR TITLE
Added consensus state to graphql interface

### PR DESCRIPTION
**Pull request by wu-s-john**
Extra fields have been added and you can check them out on this file:

Screenshot 2019-10-09 16 19 50

Checklist:

- [ ] Tests were added for the new behavior
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them:

Closes #0000
Closes #0000

### CODE REVIEW

File: src/lib/coda_graphql/coda_graphql.ml [hunk 1]
```diff
@@ -370,6 +370,39 @@ module Types = struct
               Stringable.Ledger_hash.to_base58_check
               @@ Staged_ledger_hash.ledger_hash staged_ledger_hash ) ] )
 
+  let natural_number_desc field_name num_bits =
```
Comment (bkase) [inline src/lib/coda_graphql/coda_graphql.ml]: can you call this `uint32` and put it next to the `uint64` at the top of the file and provide a similar explanation as the uint64?

File: src/lib/coda_graphql/coda_graphql.ml [hunk 2]
```diff
@@ -370,6 +370,39 @@ module Types = struct
...
+      !"String encoded natural number of the %s (%i bits)"
+      field_name num_bits
+
+  let consensus_state =
+    let open Consensus.Data.Consensus_state in
+    obj "ConsensusState" ~fields:(fun _ ->
+        [ field "blockchainLength" ~typ:(non_null string)
+            ~doc:
+              (natural_number_desc "blockchain length"
+                 Coda_numbers.Length.length_in_bits)
```
Comment (bkase) [inline src/lib/coda_graphql/coda_graphql.ml]: you can use uint32 for this too as length is a uint32 as well (we only have it in a separate module for type-safety reasons, not to override the bit length)

File: src/lib/coda_graphql/coda_graphql.ml [hunk 3]
```diff
@@ -370,6 +370,39 @@ module Types = struct
...
+            ~args:Arg.[]
+            ~resolve:(fun _ t -> Unsigned.UInt32.to_string @@ curr_epoch t)
+        ; field "epoch"
+            ~doc:(natural_number_desc "epoch in a block" 32)
+            ~typ:(non_null string)
+            ~args:Arg.[]
+            ~resolve:(fun _ t -> Unsigned.UInt32.to_string @@ curr_epoch t)
+        ; field "totalCurreny"
+            ~doc:
+              (natural_number_desc "total currency in a block"
```
Comment (bkase) [inline src/lib/coda_graphql/coda_graphql.ml]: This is `total currency` total in the whole network, not total in one block

File: src/lib/coda_graphql/coda_graphql.ml [hunk 4]
```diff
@@ -370,6 +370,39 @@ module Types = struct
...
+                 Amount.length_in_bits)
```
Comment (bkase) [inline src/lib/coda_graphql/coda_graphql.ml]: this is a uint64

File: src/lib/coda_graphql/coda_graphql.ml [hunk 5]
```diff
@@ -370,6 +370,39 @@ module Types = struct
...
+            ~doc:(natural_number_desc "slot in a block" 32)
+            ~typ:(non_null string)
+            ~args:Arg.[]
+            ~resolve:(fun _ t -> Unsigned.UInt32.to_string @@ curr_epoch t)
+        ; field "epoch"
+            ~doc:(natural_number_desc "epoch in a block" 32)
+            ~typ:(non_null string)
+            ~args:Arg.[]
+            ~resolve:(fun _ t -> Unsigned.UInt32.to_string @@ curr_epoch t)
+        ; field "totalCurreny"
```
Comment (bkase) [inline src/lib/coda_graphql/coda_graphql.ml]: typo: should be `totalCurrency`

File: src/lib/coda_graphql/coda_graphql.ml (hunk 1 above)
Comment (bkase) [reply src/lib/coda_graphql/coda_graphql.ml]: and then all these can use this in the `typ` field

File: src/lib/coda_graphql/coda_graphql.ml [hunk 6]
```diff
@@ -370,6 +370,39 @@ module Types = struct
...
+  let natural_number_desc field_name num_bits =
+    Core.sprintf
+      !"String encoded natural number of the %s (%i bits)"
+      field_name num_bits
+
+  let consensus_state =
+    let open Consensus.Data.Consensus_state in
+    obj "ConsensusState" ~fields:(fun _ ->
+        [ field "blockchainLength" ~typ:(non_null string)
+            ~doc:
```
Comment (bkase) [inline src/lib/coda_graphql/coda_graphql.ml]: can we make the doc for these more descriptive? Assuming you move the number stuff into `typ`, for this we can say:

```
blockchainLength: "Length of the blockchain at this block"
slot: "Slot in which this block was created"
epoch: "Epoch in which this block was created"
totalCurrency: "Total currency in circulation at this block"
```

File: src/lib/coda_graphql/coda_graphql.ml [hunk 7]
```diff
@@ -379,10 +412,19 @@ module Types = struct
             ~resolve:(fun _ t ->
               Stringable.State_hash.to_base58_check t.previous_state_hash )
         ; field "blockchainState"
-            ~doc:"State related to the succinct blockchain"
+            ~doc:
```
Comment (bkase) [inline src/lib/coda_graphql/coda_graphql.ml]: Alternatively: "State which is agnostic of a particular consensus algorithm"

Comment (bkase) [reply src/lib/coda_graphql/coda_graphql.ml]: They'll see the details when they look deeper, also keeps it more open-ended

File: src/lib/coda_graphql/coda_graphql.ml (hunk 5 above)
Comment (yourbuddyconner) [inline src/lib/coda_graphql/coda_graphql.ml]: totalCurreny -> `totalCurrency`

File: src/lib/coda_graphql/coda_graphql.ml [hunk 8]
```diff
@@ -379,10 +412,19 @@ module Types = struct
...
+              "Ledger merkle roots and the creation timestamp related to a \
+               block"
             ~typ:(non_null blockchain_state)
             ~args:Arg.[]
-            ~resolve:(fun _ t -> t.blockchain_state) ] )
+            ~resolve:(fun _ t -> t.blockchain_state)
+        ; field "consensusState"
+            ~doc:
+              "State used for using the consensus algorithm, Oroburous, to \
```
Comment (bkase) [inline src/lib/coda_graphql/coda_graphql.ml]: nit: "State specific to the Codaboros Proof of Stake consensus algorithm"

File: src/lib/currency/currency.ml [hunk 9]
```diff
@@ -34,6 +34,8 @@ module type Basic = sig
 
   val fold : t -> bool Triple.t Fold.t
 
+  val length_in_bits : int
```
Comment (bkase) [inline src/lib/currency/currency.ml]: therefore we can remove this

File: src/lib/currency/currency.mli [hunk 10]
```diff
@@ -30,6 +30,8 @@ module type Basic = sig
 
   val fold : t -> bool Triple.t Fold.t
 
+  val length_in_bits : int
```
Comment (bkase) [inline src/lib/currency/currency.mli]: and this

File: src/lib/coda_graphql/coda_graphql.ml (hunk 3 above)
Comment (wu-s-john) [reply src/lib/coda_graphql/coda_graphql.ml]: techically, it's the `total currency` of the whole network at the instance of the block

File: src/lib/coda_graphql/coda_graphql.ml [hunk 11]
```diff
@@ -370,6 +380,33 @@ module Types = struct
               Stringable.Ledger_hash.to_base58_check
               @@ Staged_ledger_hash.ledger_hash staged_ledger_hash ) ] )
 
+  let natural_number_desc field_name num_bits =
```
Comment (bkase) [inline src/lib/coda_graphql/coda_graphql.ml]: can you remove this function now?

File: src/lib/coda_numbers/nat.mli [hunk 12]
```diff
@@ -1,8 +1,4 @@
-module type S_unchecked = Intf.S_unchecked
-
-module type S_checked = Intf.S_checked
-
-module type S = Intf.S
```
Comment (bkase) [inline src/lib/coda_numbers/nat.mli]: we did intentionally hide the coercion to and from uint32 and uint64 so nothing bad could happen, but we do need to expose it to access the data from graphql, so this seems fine

**Pull request merged**